src/ensembl/src/shared/components/pointer-box/PointerBox.tsx

```diff
@@ -0,0 +1,205 @@
+  container?: HTMLElement | null; // area within which the box should try to position itself; defaults to window if null
+  autoAdjust: boolean; // whether to adjust pointer box position so as not to extend beyond screen bounds
+  renderInsideAnchor: boolean; // whether to render PointerBox inside the anchor (which should have position: relative to display it properly); renders to body if false
+  pointerWidth: number;
+  pointerHeight: number;
+  pointerOffset: number;
+  classNames?: {
+    body?: string;
+    pointer?: string;
+  };
+  children: ReactNode;
+  onMouseEnter: () => void;
+  onMouseLeave: () => void;
+  onOutsideClick: () => void;
+  onClose: () => void;
+};
+
+const handleClickInside = (e: React.MouseEvent | React.TouchEvent) => {
+  e.preventDefault();
+  e.stopPropagation(); // this works within React's event system
+  e.nativeEvent.stopImmediatePropagation(); // also prevent propagation to DOM elements outside of React (e.g. to document)
+};
+
+const PointerBox = (props: PointerBoxProps) => {
+  const [isPositioning, setIsPositioning] = useState(props.autoAdjust);
+  const positionRef = useRef<Position | null>(props.position);
+  const anchorRectRef = useRef<DOMRect | null>(null);
+  const [inlineStyles, setInlineStyles] = useState<InlineStylesState>({
+    bodyStyles: {},
+    pointerStyles: {}
+  });
+  const pointerBoxRef = useRef<HTMLDivElement>(null);
+
+  useOutsideClick(pointerBoxRef, props.onOutsideClick);
+
+  useEffect(() => {
+    const pointerBoxElement = pointerBoxRef.current as HTMLDivElement;
+
+    setInlineStyles(getInlineStyles(props));
+    anchorRectRef.current = props.anchor.getBoundingClientRect();
+
+    if (props.autoAdjust) {
+      adjustPosition(pointerBoxElement, props.anchor);
+    }
+  }, []);
+
+  const closeOnScroll = once(props.onClose);
+
+  // from Stack Overflow: https://stackoverflow.com/questions/59792071/how-to-observe-dom-element-position-changes
+  // (listen to all scroll events in the event capturing phase on the body and re-evaluate anchor position)
+  useEffect(() => {
+    if (!props.renderInsideAnchor) {
+      document.addEventListener('scroll', closeOnScroll, true);
+      return () => document.removeEventListener('scroll', closeOnScroll, true);
+    }
+  }, []);
+
+  const getInlineStyles = (props: PointerBoxProps) => {
+    if (props.renderInsideAnchor) {
+      return getStylesForRenderingIntoAnchor(props);
+    } else {
+      return getStylesForRenderingIntoBody(props);
+    }
+  };
+
+  const adjustPosition = (pointerBox: HTMLDivElement, anchor: HTMLElement) => {
+    const pointerBoxBoundingRect = pointerBox.getBoundingClientRect();
+    const rootBoundingRect = props.container
+      ? props.container.getBoundingClientRect()
+      : windowService.getDimensions();
+    const anchorBoundingRect = anchor.getBoundingClientRect();
+
+    const optimalPosition = findOptimalPosition({
+      pointerBoxBoundingRect,
+      anchorBoundingRect,
+      rootBoundingRect,
```

Comment:
When we render the toolbox inside an anchor, are we doing anything by calculating `rootBoundingRect`?
![Screenshot 2020-03-25 at 07 34 16](https://user-images.githubusercontent.com/3085815/77512917-d9a13c00-6e6b-11ea-90f3-28c01e403ebd.png)

Reply:
Looking at the code. The rootBoundingRect is used to get dimensions of the container if the pointerBox has to be aware of its borders

![image](https://user-images.githubusercontent.com/6834224/77527033-61467500-6e83-11ea-9da8-7b1263d9e453.png)